File: benchmarks-haskell/scott-merge-sort.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}

{-# HLINT ignore "Use const" #-}
{-# HLINT ignore "Redundant bracket" #-}

newtype ScottList a = ScottList (forall r. r -> (a -> ScottList a -> r) -> r)

unlist :: ScottList a -> forall r. r -> (a -> ScottList a -> r) -> r
unlist (ScottList f) = f

scottNil :: ScottList a
scottNil = ScottList (\n c -> n)

scottCons :: a -> ScottList a -> ScottList a
scottCons h t = ScottList (\n c -> c h t)

scottSingleton :: a -> ScottList a
scottSingleton x = scottCons x scottNil

scottSplit :: ScottList a -> (forall r. (ScottList a -> ScottList a -> r) -> r)
scottSplit list k =
  (unlist list)
    (k scottNil scottNil)
    ( \x xs ->
        (unlist xs)
          (k (scottSingleton x) scottNil)
          ( \y ys ->
              scottSplit
                ys
                (\left right -> k (scottCons x left) (scottCons y right))
          )
    )

scottMerge :: ScottList Int -> ScottList Int -> ScottList Int
scottMerge xs ys =
  (unlist ys)
    xs
    ( \y yss ->
        (unlist xs)
          ys
          ( \x xss ->
              if x < y
                then scottCons x (scottMerge xss ys)
                else scottCons y (scottMerge xs yss)
          )
    )

scottMergeSort :: ScottList Int -> ScottList Int
scottMergeSort list =
  (unlist list)
    scottNil
    ( \x xs ->
        (unlist xs)
          (scottSingleton x)
          ( \dummy dummyx ->
              scottSplit
                list
                ( \left right ->
                    scottMerge (scottMergeSort left) (scottMergeSort right)
                )
          )
    )

scottSumList :: ScottList Int -> Int
scottSumList list = (unlist list) 0 (\x xs -> x + scottSumList xs)

generateList :: Int -> ScottList Int
generateList n = go 0 scottNil
  where
    go i acc
      | i < n = go (i + 1) (scottCons i acc)
      | otherwise = acc

main :: IO ()
main =
  print $ scottSumList $ scottMergeSort $ generateList 1000

File: benchmarks-ocaml/scott_merge_sort.ml
type 'a scott_list = { unlist : 'r. 'r -> ('a -> 'a scott_list -> 'r) -> 'r }

let scott_nil = { unlist = (fun n c -> n) }

let scott_cons h t = { unlist = (fun n c -> c h t) }

let scott_singleton x = scott_cons x scott_nil

let rec scott_split list k =
  list.unlist (k scott_nil scott_nil) (fun x xs ->
      xs.unlist
        (k (scott_singleton x) scott_nil)
        (fun y ys ->
          scott_split ys (fun left right ->
              k (scott_cons x left) (scott_cons y right))))

let rec scott_merge xs ys =
  ys.unlist xs (fun y yss ->
      xs.unlist ys (fun x xss ->
          if x < y then scott_cons x (scott_merge xss ys)
          else scott_cons y (scott_merge xs yss)))

let rec scott_merge_sort list =
  list.unlist scott_nil (fun x xs ->
      xs.unlist (scott_singleton x) (fun dummy dummyx ->
          scott_split list (fun left right ->
              scott_merge (scott_merge_sort left) (scott_merge_sort right))))

let rec scott_sum_list list = list.unlist 0 (fun x xs -> x + scott_sum_list xs)

let generate_list n =
  let rec go i acc = if i < n then go (i + 1) (scott_cons i acc) else acc in
  go 0 scott_nil

let () =
  print_int (scott_sum_list (scott_merge_sort (generate_list 100)));
  print_newline ()
